### PR TITLE
PCHR-3180: Do not recalculate balance on cancelling or rejecting from Leave Request Modal

### PR DESCRIPTION
**Pull request by igorpavlov**
## Overview

When the work pattern has changed so it affects the balance change of a leave request, and such a leave request gets cancelled/rejected via Leave Request modal the balance change recalculation prompt appears.

## Technical Details

```js
// new function isBalanceChangeRecalculationNeeded()
function isBalanceChangeRecalculationNeeded () {
  return !vm.request.status_id || !_.includes(['cancelled', 'rejected'], getStatusFromValue(vm.request.status_id).name);
}

function submit () {
  // ...
  return vm.request.isValid()
    .then(isBalanceChangeRecalculationNeeded() && checkIfBalanceChangeHasChanged)

function decideIfBalanceChangeNeedsRecalculationOnBackend () {
  if (isBalanceChangeRecalculationNeeded() && !vm.isRole('staff')) {
    vm.request.change_balance = true;
```
## Comments
https://github.com/civicrm/civihr/pull/2521/commits/e6137e18f97aabf13044a054da2be935fee74478

The problem is that backend will recalculate the balance if dates are changed disregarding if you pass `change_balance` or not. This is due to:

```php
private function skipBalanceChangeUpdate($params) {
    // ...
    $leaveRequestDatesNotChanged = !$isToilRequest && $datesDidNotChange; // <<<
    // ...
    $updateBalanceChange = !empty($params['change_balance']); // <<<

    return ($leaveRequestDatesNotChanged || $toilDatesAndToAccrueNotChanged) && !$updateBalanceChange;
  }
```
and when the work pattern day changes, let's say, to a weekend, then the time in leave request modal is set to 00:00. Please note, if you *approve* a request, then it needs to save new time (00:00). If you cancel after changing request dates, times or deductions, then it also needs to ave new time. But if you do not touch leave request and just wish to set status to Rejected or Cancel, then it needs to store the original time and do not recalculate balance on backend.

Additionally, this commit https://github.com/civicrm/civihr/pull/2521/commits/e6137e18f97aabf13044a054da2be935fee74478 introduces a change that reverts leave request time to the original value if the request has not been touched. Here is an example of the situation when it is needed:

1) If you create a leave request in hours let's say for 9:00 - 14:00
2) Then work pattern changes to 10:00 - 18:00 (meaning 9:00 is not a valid time anymore)
3) When you open the leave request you will see 10:00 - 14:00, that is expected
4) If you **cancel** the request the time should stay as 09:00.

```js
function submit () {
  // ...
  return vm.request.isValid()
    // ...
    .then(!vm.request.change_balance && revertRequestOriginalDatesAndTimes)
    // ...

function revertRequestOriginalDatesAndTimes () {
  ['from', 'to'].forEach(function (dateType) {
    vm.request[dateType + '_date'] =
      initialLeaveRequestAttributes[dateType + '_date'];
  });
}
```

------
✅Manual Tests - passed
✅Jasmine Tests - passed
✅JS distributive files - included
⏹Backstop tests - not needed
⏹CSS distributive files - not needed
⏹PHP Unit Tests - not needed

### CODE REVIEW

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js [hunk 1]
```diff
@@ -297,8 +297,8 @@ define([
      * Sets the "change_balance" attribute to the leave request
      * if a force balance change recalculation is needed
      */
-    function decideIfBalanceChangeNeedsAForceRecalculation () {
-      if (!vm.isRole('staff')) {
+    function decideIfBalanceChangeNeedsRecalculationOnBackend () {
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js]: `isBalanceChangeNeedsRecalculationOnBackend` ? Does it sound better(not sure)

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js [hunk 2]
```diff
@@ -297,8 +297,8 @@ define([
...
+      if (isBalanceChangeRecalculationNeeded() && !vm.isRole('staff')) {
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js]: I want to understand how this `change_balance` property is fetched. Because it will depend on the status which you set before saving, so do we get that while calling `isValid()`?

Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js]: Are we removing the `change_balance` property before sending it back to API(`toAPI`), because its not needed by backend.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/request.controller.spec.js [hunk 3]
```diff
@@ -705,38 +704,44 @@
...
+                expect(controller.request.update).not.toHaveBeenCalled();
+              });
+
+              it('prompts a balance change recalculation', function () {
+                expect(LeaveRequestService.promptIfProceedWithBalanceChangeRecalculation)
+                  .toHaveBeenCalled();
               });
 
-              describe('after recalculation on submit attempt', function () {
+              describe('on confirm the balance change recalculation', function () {
```
[the remark this answers is not in the document]
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/request.controller.spec.js]: Done.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/request.controller.spec.js [hunk 4]
```diff
@@ -705,38 +704,44 @@
...
 
-              it('initiates the balance change recalculation', function () {
-                expect($rootScope.$emit).toHaveBeenCalledWith(
-                  'LeaveRequestPopup::recalculateBalanceChange');
+              it('does not call update method on instance', function () {
+                expect(controller.request.update).not.toHaveBeenCalled();
+              });
+
+              it('prompts a balance change recalculation', function () {
+                expect(LeaveRequestService.promptIfProceedWithBalanceChangeRecalculation)
```
[the remark this answers is not in the document]
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/request.controller.spec.js]: `confirmBalanceChangeRecalculation` is doing different thing - but we can change to `promptBalanceChangeRecalculation`

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js (hunk 1 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js]: `isBalanceChangeRecalculationNeededOnBackend`

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js (hunk 2 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js]: It **is** needed for backend, it is essential actually, since then backend decided if to skip recalculation in case of dates/times have not been changed.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js [hunk 5]
```diff
@@ -909,8 +919,8 @@ define([
       amendRequestParamsBeforeSave();
 
       return vm.request.isValid()
-        .then(checkIfBalanceChangeHasChanged)
-        .then(decideIfBalanceChangeNeedsAForceRecalculation)
+        .then(isBalanceChangeRecalculationNeeded() && checkIfBalanceChangeHasChanged)
+        .then(isBalanceChangeRecalculationNeededOnBackend)
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js]: Really sorry, but the previous name was better, because `isBalanceChangeRecalculationNeededOnBackend` sounds like it is going to return a boolean value. Please revert this change.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js [hunk 6]
```diff
@@ -614,6 +604,26 @@ define([
...
+     *
+     * @return {Boolean}
+     */
+    function isBalanceChangeRecalculationNeeded () {
+      return !vm.request.status_id || !_.includes(['cancelled', 'rejected'], getStatusFromValue(vm.request.status_id).name);
+    }
+
+    /**
+     * Sets the "change_balance" attribute to the leave request
+     * if a force balance change recalculation is needed
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js]: `is needed on the backend`

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js (hunk 2 above)
Comment (deb1990) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js]: Ohh, my bad.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js [hunk 7]
```diff
@@ -826,6 +829,27 @@ define([
       vm.dismissModal();
     }
 
+    /**
+     * Stores the request from and to times original values
+     */
+    function rememberRequestOriginalTimes () {
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js]: I dont think we need this new variable and function. We already have `initialLeaveRequestAttributes` which stores the request object's original state. So we can use the same.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js [hunk 8]
```diff
@@ -826,6 +829,27 @@ define([
...
+      ['from', 'to'].forEach(function (dateType) {
+        originalRequestTimes[dateType] =
+          moment(vm.request[dateType + '_date']).format('HH:mm');
+      });
+    }
+
+    /**
+     * Reverts the request from and to times back to the original values
+     */
+    function revertRequestOriginalTimes () {
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js]: Also the PR is not updated with these changes of this commit. Please add them.

Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js]: Done.
